ehsm_kms_service/secret_manager_apis.js

```diff
@@ -265,9 +367,678 @@ const createSecret = async (res, appid, payload, DB) => {
+ * @returns
+ */
+const putSecretValue = async (res, appid, DB, payload) => {
+    try {
+        //get and check Param in payload
+        let secretName = getParam_String(payload, 'secretName')
+        let secretData = getParam_String(payload, 'secretData')
+        const createTime = new Date().getTime()
+        let sm_masterKey
+        let encryptionKeyId
+        let versionId
+        if (!checkStringParam(secretData, true, SECRETDATA_LENGTH_MAX)) {
+            res.send(_result(400, `secretData cannot be empty, must be string and length not more than ${SECRETDATA_LENGTH_MAX}`))
+            return
+        }
+        if (!checkStringParam(secretName, true, SECRETNAME_LENGTH_MAX)) {
+            res.send(_result(400, `secretName cannot be empty, must be string and length not more than ${SECRETNAME_LENGTH_MAX}`))
+            return
+        }
+
+        //query SM_Masterkey and encrypt secretData
+        const query_secertName = {
+            selector: {
+                appid,
+                secretName,
+            },
+            limit: 1
+        }
+        const secret_metadata_res = await DB.partitionedFind('secret_metadata', query_secertName)
+        if (secret_metadata_res.docs.length == 0) {
+            res.send(_result(400, 'can not find secretName'))
+            return
+        } else {
+            if (secret_metadata_res.docs[0].encryptionKeyId == undefined) {
+                encryptionKeyId = ""
+            }
+            else {
+                encryptionKeyId = secret_metadata_res.docs[0].encryptionKeyId
+            }
+        }
+        retCMK = await get_SM_Masterkey(DB, appid, encryptionKeyId)
+        if (retCMK.isFind) {
+            sm_masterKey = retCMK.sm_masterKey
+        } else {
+            res.send(_result(400, retCMK.msg))
+            return
+        }
+        const apikey_encrypt_res = napi_result(cryptographic_apis.Encrypt, res, [sm_masterKey, secretData, ''])
+        // check encrypt status and get ciphertext
+        if (!apikey_encrypt_res) {
+            return
+        }
+        if (apikey_encrypt_res.code != 200) {
+            res.send(apikey_encrypt_res)
+            return
+        }
+        const { ciphertext } = apikey_encrypt_res.result
+
+        //Search the old version by appid, secretName, versionStage, deletedFlag 
+        //change old version stage and add a new version
+        const query_version_stage = {
+            selector: {
+                appid,
+                secretName,
+                versionStage: SM_SECRET_VERSION_STAGE_CURRENT,
+            },
+            fields: [
+                '_id',
+                '_rev',
+                'appid',
+                'secretName',
+                'versionId',
+                'secretData',
+                'createTime',
+                'versionStage',
+            ],
+            limit: 1
+        }
+        const version_stage_res = await DB.partitionedFind('secret_version_data', query_version_stage)
```

Comment:
how do you check whether the secret is logically deleted or not now?

Reply:
deleteTime: { "$eq": null }

Reply:
if deleteTime  != null the secret is logically deleted, if deleteTime  == null the secret is not logically delete.

Reply:
Does putSecretValue check  deleteTime: { "$eq": null } ?

Reply:
And updateSecretDesc? others?

Reply:
The putSecretValue will add deleteTime verification, Other api no need.